src/types/iterator.rs

```diff
@@ -56,18 +56,60 @@ impl<'p> Iterator for &'p PyIterator {
     /// Further `next()` calls after an exception occurs are likely
     /// to repeatedly result in the same exception.
     fn next(&mut self) -> Option<Self::Item> {
-        let py = self.0.py();
+        Bound::borrowed_from_gil_ref(self)
+            .next()
+            .map(|result| result.map(Bound::into_gil_ref))
+    }
+
+    #[cfg(not(Py_LIMITED_API))]
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        Bound::borrowed_from_gil_ref(self).size_hint()
+    }
+}
+
+impl<'py> Iterator for Bound<'py, PyIterator> {
+    type Item = PyResult<Bound<'py, PyAny>>;
+
+    /// Retrieves the next item from an iterator.
+    ///
+    /// Returns `None` when the iterator is exhausted.
+    /// If an exception occurs, returns `Some(Err(..))`.
+    /// Further `next()` calls after an exception occurs are likely
+    /// to repeatedly result in the same exception.
+    #[inline]
+    fn next(&mut self) -> Option<Self::Item> {
+        (&mut &*self).next()
+    }
 
-        match unsafe { py.from_owned_ptr_or_opt(ffi::PyIter_Next(self.0.as_ptr())) } {
+    #[cfg(not(Py_LIMITED_API))]
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let hint = unsafe { ffi::PyObject_LengthHint(self.as_ptr(), 0) };
+        (hint.max(0) as usize, None)
+    }
+}
+
+impl<'py> Iterator for &'_ Bound<'py, PyIterator> {
```

Comment:
Could you explain why this impl is needed? I would have assumed that std's [`by_ref`](https://doc.rust-lang.org/stable/std/iter/trait.Iterator.html#method.by_ref) method would be sufficient? (I understand a mutable reference is strange in Python land, it is a mutable reference to pointer after all which should straight forward to produce even in a PyO3 program (at worst by cloning an existing `&Bound`).

Reply:
I feel like it was convenient in some context when I was experimenting with this API in the past (which is why it's ended up here), but I can't recall the concrete example right now so let's delete it and can add it back later if we find motivation.